### PR TITLE
Macaroon v2

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

This PR starts the work to eliminate the use of macaroon.v2-unstable and bakerty.v2-unstable upstream repos. The work completed so far includes the use of macaroon.v2 everywhere and the use of bakery.v2 in the client (essentially bakery.v2/httpbakery).

The juju controller still uses bakery.v2-unstable to mint and store macaroons, and perform authorisation checks. The v2 APIs are totally different to v2-unstable so major work will be needed to get that fixed.

Much of the PR is the import changes to replace macaroon.v2-unstable with macaroon.v2 (and the similar bakery.v2 import changes).

There were several side effects that had to be addressed. We needed to move to using charmrepo.v4 instead of v3 in order to be compatible with the v2 macaroons etc. The v4 API changes had a flow on effect to the cmd/juju/application tests. The good part is that a huge chunk of code can be deleted (fakecharmstore) and all the deploy/bundle tests are much more consistent with each other.

Another side effect is that some code used juju/httprequest and other bits used httprequest.v1. Some work was needed to consolidate on httprequest.v1 and eliminate upstream conflicts.

To enable the use of macaroon.v2 everywhere, we need to depend on a fork I did of the bakery.v2-unstable repo. The fork introduced macaroon.v2 instead of v2-unstable to the bakery and did work to fix the juju/httprequest vw httprequest.v1 conflicts. As soon as we move to bakery.v2 everywhere, this temp dependency will be removed.

The end result is that all juju client code - CLI, api client, controller access to charm store etc - uses bakery.v2 and macaroon.v2 everywhere. Only the controller bakery and macaroon checks remain on bakery.v2 unstable.

## QA steps

We need to test old/new clients talking to new/old controllers, and also jaas.

Using a new juju client:
- login to jaas
- logout of jaas
- clear cookies from .local/share/juju
- log in to jaas again (both with -B and without -B to test interactive and browser auth)
- log in to an old 2.7 juju controller
- logout and see the change password prompt
- change the password, logout, and log in again
- add a user and use a different client home dir to register that user

Using an old 2.7 juju client
- perform the above checks against a new juju controller

Test CMR macaroon auth by:
- bootstrapping an old and a new controller
- in the old controller, create an offer
- add a user, grant that user access to the offer
- consume the offer in the new controller as the non admin user
- rinse and repeat where the offer is in the new controller

### CODE REVIEW

File: api/client_macaroon_test.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ func (s *clientMacaroonSuite) TestAddLocalCharmWithFailedDischarge(c *gc.C) {
 		fmt.Sprintf("local:quantal/%s-%d", charmArchive.Meta().Name, charmArchive.Revision()),
 	)
 	savedURL, err := client.AddLocalCharm(curl, charmArchive, false)
-	c.Assert(err, gc.ErrorMatches, `POST https://.+: cannot get discharge from "https://.*": third party refused discharge: cannot discharge: login denied by discharger`)
+	c.Assert(err, gc.ErrorMatches, `Post https://.+: cannot get discharge from "https://.*": third party refused discharge: cannot discharge: login denied by discharger`)
```
Comment (hpidcock) [inline api/client_macaroon_test.go]: Interested as to why this changed :/

Comment (wallyworld) [reply api/client_macaroon_test.go]: This is due to a change in the upstream httprequest library (from juju/httprequest to httprequest.v1)

File: featuretests/bakerystorage_test.go [hunk 2]
```diff
@@ -11,14 +11,14 @@ import (
...
-	"gopkg.in/macaroon.v2-unstable"
+	"gopkg.in/macaroon.v2"
 	"gopkg.in/mgo.v2"
 
 	"github.com/juju/juju/mongo"
 	"github.com/juju/juju/state/bakerystorage"
 )
 
-// This suite is not about a feature tests per se, but tests the integration
+// This suite is not about a feature tests per se, but tests the integrationmgostorage
```
Comment (hpidcock) [inline featuretests/bakerystorage_test.go]: integrationmgostorage?

File: apiserver/common/crossmodel/auth.go [hunk 3]
```diff
@@ -12,12 +12,14 @@ import (
 	"gopkg.in/juju/names.v3"
 	"gopkg.in/macaroon-bakery.v2-unstable/bakery"
 	"gopkg.in/macaroon-bakery.v2-unstable/bakery/checkers"
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/common/crossmodel/auth.go]: explained in the PR description :-)

File: api/http_test.go [hunk 4]
```diff
@@ -256,7 +257,7 @@ func (s *httpSuite) TestAuthHTTPRequest(c *gc.C) {
 }
 
 func (s *httpSuite) authHTTPRequest(c *gc.C, info *api.Info) *http.Request {
-	req, err := http.NewRequest("GET", "/", nil)
+	req, err := http.NewRequest("Get", "/", nil)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/http_test.go]: uses const now

File: featuretests/bakerystorage_test.go (hunk 2 above)
Comment (wallyworld) [reply featuretests/bakerystorage_test.go]: cut/paste error, fixed

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

File: api/authentication/interactor.go [hunk 5]
```diff
@@ -4,47 +4,53 @@
...
 	"github.com/juju/errors"
-	"gopkg.in/macaroon-bakery.v2-unstable/httpbakery"
+	schemaform "gopkg.in/juju/environschema.v1/form"
+	"gopkg.in/macaroon-bakery.v2/httpbakery"
+	"gopkg.in/macaroon-bakery.v2/httpbakery/form"
 )
 
 const authMethod = "juju_userpass"
 
 // Visitor is a httpbakery.Visitor that will login directly
```
Comment (manadart) [inline api/authentication/interactor.go]: Missed corresponding comment change here.

**Comment by achilleasa** [comment]
⚠️  charmrepo.v3 includes commits that are **not** available in charmrepo.v4 and will need to be forward-ported.

More specifically, these ones are mine: https://github.com/juju/charmstore/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+author%3Aachilleasa and implement the channel overrides in bundles feature.

The PR also reverts some of the required changes (e.g. https://github.com/juju/juju/pull/11045/files#diff-6290a8a735de16d90289b6021f12691cR377) for supporting channels in bundles that were introduced by https://github.com/juju/juju/pull/10480.

I will try to forward-port these today and open a PR to re-introduce the removed code. 

However, I think @nammn also has some commits in there that also need to be ported to make sure that we don't introduce any regressions.

**Pull request merged**